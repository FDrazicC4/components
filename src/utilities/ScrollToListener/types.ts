import { ReactNode } from 'react';


export interface Props {


    /**
     * Tells the scroll listener to apply it's dom ref to the last child instead of first child
     * Useful for infinite scrolling; Default is useful for React Fragment Components
     */
    applyToLastChild?: boolean,


    /**
     * Child element that the scroll to listener will be applied to
     */
    children: ReactNode,


    /**
     * Forwarded Ref
     */
    forwardedRef?: HTMLElement | null,


    /**
     * Function to call when the element comes into the viewport
     */
    onScrollTo: () => void,


    /**
     * Must be ancestor of target.
     * Defaults to browser viewport if not specified/null
     */
    // rootElement?: Element | null,


    /**
     * A string, formatted similarly to the CSS margin property.
     * Used to grow or shrink each side of the root element's bounding box before computing intersections
     * May use %'s
     */
    rootMargin?: string,


    /**
     * Either a number or array of numbers indicating when the callback should be executed
     * 0 or [0] or [0, 0.25, 0.50, 0.75, 1]
     */
    rootThreshold?: number[] | number,


    /**
     * Specifies if you want the callback to continuously listen when scrolling down and up
     * Default specifies to only fire callback once, when scrolling into view initially
     */
    willListen?: boolean,


}
