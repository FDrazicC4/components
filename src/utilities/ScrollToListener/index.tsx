import React, { ReactElement, RefObject } from 'react';
import { Props } from './types';


/**
 * Fires an function passed through onScrollTo when the child element is scrolled to
 */
function ScrollToListener({
    children,
    onScrollTo,
    rootElement = null,
    rootMargin = '0px 0px 0px 0px',
    rootThreshold = 0.01,
    willListen = false,
}: Props): ReactElement {


    const ref: RefObject<HTMLDivElement> = React.useRef(null);


    function handleScroll(entries: any, observer: any): void {


        const [entry] = entries;
        const lastThreshold = observer.thresholds[observer.thresholds.length - 1];


        if (entry.intersectionRatio > 0) onScrollTo();
        if (!willListen && entry.intersectionRatio >= lastThreshold) observer.disconnect();


    }


    React.useEffect(() => {


        const findRootElement = rootElement ? document.querySelector(rootElement) : null;

        const target = ref.current?.previousElementSibling;

        const observer = new IntersectionObserver(handleScroll, { root: findRootElement,
            rootMargin,
            threshold: rootThreshold });


        if (target) observer.observe(target);


        return ((): void => {
            if (target) observer.unobserve(target);
        });


    }, []);


    return (
        <React.Fragment>
            {children}
            <div ref={ref} style={{ display: 'none' }} />
        </React.Fragment>
    );


}


export default ScrollToListener;
