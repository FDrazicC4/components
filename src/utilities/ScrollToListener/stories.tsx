import React, { ReactElement } from 'react';
import ScrollToListener from './index';


export default {
    component: ScrollToListener,
    title: 'Utilities|ScrollToListener',
};


export const Default = (): ReactElement => {
    // eslint-disable-next-line no-console
    const fire = (): void => console.log('scrolled to');


    return (
        <div style={{ backgroundColor: 'rebeccapurple', height: '1000px', width: '1000px' }}>
            <ScrollToListener onScrollTo={fire}>
                <div>HELLO WORLD</div>
            </ScrollToListener>
        </div>
    );


};
