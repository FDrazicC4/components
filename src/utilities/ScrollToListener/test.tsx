import React from 'react';
import renderer from 'react-test-renderer';
import ScrollToListener from './index';


it('renders correctly', () => {
    const tree = renderer
        .create((
            <ScrollToListener onScrollTo={(): boolean => true}>
                <div>foo</div>
            </ScrollToListener>
        ))
        .toJSON();


    expect(tree).toMatchSnapshot();
});
