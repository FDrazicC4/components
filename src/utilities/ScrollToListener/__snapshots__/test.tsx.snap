// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <span
    style={
      Object {
        "display": "none",
      }
    }
  />,
  <div>
    foo
  </div>,
]
`;
